cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py

```diff
@@ -29,15 +29,44 @@
     import cirq
 
 
-def _create_transformer_with_kwargs(func: 'cirq.TRANSFORMER', **kwargs) -> 'cirq.TRANSFORMER':
-    """Hack to capture additional keyword arguments to transformers while preserving mypy type."""
+def create_transformer_with_kwargs(transformer: 'cirq.TRANSFORMER', **kwargs) -> 'cirq.TRANSFORMER':
+    """Method to capture additional keyword arguments to transformers while preserving mypy type.
```

Comment:
I worry that by silencing mypy's objection we might be hiding an underlying issue in our API.

IIUC the issue is that "context" and other keyward arguments are lumped together by the API even though they seem to pertain to different actions and different stages in the life of a transformer: context is needed when a transformer is acting on a circuit and - AFAICT - the other keyword arguments are needed at transfromer instantiation time. (The issue I described in the other comment below is also a symptom of the loss of this distinction.)

Of course this might be down to the switch from transformer classes to transformer functions. If this is the case, then I wonder whether we should use a context manager here. WDYT?

Reply:
Transformer classes are still supported; so a user can create a new transformer by decorating a transformer class with `@cirq.transformer` decorator, and use an instance of that class everywhere an API expects a `cirq.TRANSFORMER`. 

Many transformers are now implemented as functions and not classes because the transformers implemented as functions are stateless and the only processing happening between transformer instantiation and transformer acting on the circuit is storing the init  **kwargs arguments as members of the class using them when the transformer is applied on the circuit. In this case, forcing users to create transformers as classes only to store and reuse **kwargs was too much boilerplate, and therefore we decided to support both functions and classes as part of the API. 

This convenience method will enable us to bind known **kwargs to transformers in advance before passing them to APIs that expect a `cirq.TRANSFORMER` and are not aware of additional **kwargs. 

IIUC, context managers are helpful for managing resources via automatic setup and teardown; I'm not sure if we need the heavy machinery here. Did you have a usage in mind which only rewrites this particular method to make it more robust; or is your proposal to change the Transformer API design in some way? If it's the latter; I'd say we also don't have enough time to do any of that right before 1.0 now.

Reply:
Makes sense. This function creates a scope that holds on to two extra kwargs without having to declare a new class for each transformer.